File: mobile/native/OctoTunnelPlugin.swift
import Capacitor
import CryptoKit
import Foundation
import Security

// OctoTunnel is the phone half of octo's managed tunnel — the native piece that
// cannot live in JavaScript. It holds the device's Noise static keypair in the
// iOS Keychain, runs the Noise XX handshake with the host (brokered by the
// relay, which sees only ciphertext), and moves encrypted frames over a relay
// WebSocket. JavaScript hands it plaintext frames and receives plaintext frames
// back; no key material ever crosses the bridge.
//
// It mirrors the phone (initiator) side of the Go reference client in
// cmd/octo-relay/internal/client and the Kotlin plugin, and interoperates with
// the host in internal/tunnel: Noise_XX_25519_ChaChaPoly_SHA256, empty prologue,
// JSON relay frames {t, tn, d, p} with p a base64 payload the relay never reads.
@objc(OctoTunnelPlugin)
public class OctoTunnelPlugin: CAPPlugin, CAPBridgedPlugin {
    // Capacitor plugin identity + method table, declared in Swift (the modern
    // pure-Swift plugin form) so no companion Objective-C CAP_PLUGIN macro file
    // is needed. jsName is the name the JS side calls: registerPlugin('OctoTunnel').
    public let identifier = "OctoTunnelPlugin"
    public let jsName = "OctoTunnel"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "loadDeviceIdentity", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pair", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "send", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "disconnect", returnType: CAPPluginReturnPromise),
    ]

    private let queue = DispatchQueue(label: "dev.octo.tunnel")
    private var handshake: NoiseHandshakeXX?
    private var tx: NoiseCipherState?
    private var rx: NoiseCipherState?
    private var ws: URLSessionWebSocketTask?
    private var urlSession: URLSession?
    private var deviceId: String?
    private var expectedHostKey: Data?
    private var pairCall: CAPPluginCall?

    // MARK: - JS contract (native-bridge.ts)

    @objc func loadDeviceIdentity(_ call: CAPPluginCall) {
        do {
            let identity = try deviceIdentity()
            call.resolve(["publicKey": identity.publicKey.base64EncodedString()])
        } catch {
            call.reject("load identity failed: \(error.localizedDescription)")
        }
    }

    @objc func pair(_ call: CAPPluginCall) {
        guard let relay = call.getString("relay"), !relay.isEmpty,
              let token = call.getString("token"), !token.isEmpty,
              let hostKey = call.getString("hostKey"), !hostKey.isEmpty else {
            call.reject("pair: relay, token and hostKey are required")
            return
        }
        guard let expected = Data(base64Encoded: hostKey) else {
            call.reject("pair: hostKey is not valid base64")
            return
        }
        queue.async {
            do {
                let identity = try self.deviceIdentity()
                self.handshake = try NoiseHandshakeXX(staticPrivate: identity.privateKey)
                self.tx = nil
                self.rx = nil
                self.deviceId = nil
                self.expectedHostKey = expected
                self.pairCall = call

                // URLSessionWebSocketTask dials ws(s):// directly (unlike OkHttp,
                // which upgrades over http). The PoC relay routes by token;
                // production dials <tunnelId>.<relay-host> for SNI routing.
                let base = relay.hasSuffix("/") ? String(relay.dropLast()) : relay
                guard let url = URL(string: "\(base)/device?token=\(token)") else {
                    throw TunnelError.message("bad relay URL")
                }
                let session = URLSession(configuration: .default)
                let task = session.webSocketTask(with: url)
                self.urlSession = session
                self.ws = task
                task.resume()
                self.receiveLoop()
            } catch {
                self.pairCall = nil
                call.reject("pair failed: \(error.localizedDescription)")
            }
        }
    }

    @objc func send(_ call: CAPPluginCall) {
        guard let frame = call.getString("frame") else {
            call.reject("send: frame is required")
            return
        }
        queue.async {
            guard let tx = self.tx else {
                call.reject("send: no active session")
                return
            }
            do {
                let ct = try tx.encrypt(Data(frame.utf8))
                self.sendFrame(type: "data", payload: ct)
                call.resolve()
            } catch {
                call.reject("send failed: \(error.localizedDescription)")
            }
        }
    }

    @objc func disconnect(_ call: CAPPluginCall) {
        queue.async {
            self.teardown(nil)
            call.resolve()
        }
    }

    // MARK: - relay WebSocket

    private func receiveLoop() {
        ws?.receive { [weak self] result in
            guard let self = self else { return }
            switch result {
            case .failure(let error):
                self.queue.async { self.teardown(error) }
            case .success(let message):
                let text: String
                switch message {
                case .string(let s): text = s
                case .data(let d): text = String(decoding: d, as: UTF8.self)
                @unknown default: text = ""
                }
                self.queue.async { self.onMessage(text) }
                self.receiveLoop()
            }
        }
    }

    private func onMessage(_ text: String) {
        guard let data = text.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = obj["t"] as? String else { return }
        do {
            switch type {
            case "paired":
                deviceId = obj["d"] as? String
                // Initiator opens the XX handshake: send message 1 (e).
                if let hs = handshake {
                    let msg1 = try hs.writeMessage1()
                    sendFrame(type: "handshake", payload: msg1)
                }
            case "handshake":
                guard let hs = handshake, let p = obj["p"] as? String,
                      let msg = Data(base64Encoded: p) else { return }
                // The one inbound handshake message in XX is msg2 (e, ee, s, es).
                let msg3 = try hs.readMessage2WriteMessage3(msg)
                // Authenticate the host: its static key, learned in msg2, must
                // equal the one the pairing QR carried — the relay can't forge it.
                if let expected = expectedHostKey, hs.remoteStatic != expected {
                    teardown(TunnelError.message("host key mismatch — refusing to trust relay peer"))
                    return
                }
                sendFrame(type: "handshake", payload: msg3)
                let keys = hs.split()
                tx = NoiseCipherState(key: keys.send)
                rx = NoiseCipherState(key: keys.receive)
                handshake = nil
                pairCall?.resolve()
                pairCall = nil
            case "data":
                guard let rx = rx, let p = obj["p"] as? String,
                      let ct = Data(base64Encoded: p) else { return }
                let pt = try rx.decrypt(ct)
                let frame = String(decoding: pt, as: UTF8.self)
                notifyListeners("frame", data: ["frame": frame])
            default:
                break
            }
        } catch {
            teardown(error)
        }
    }

    private func sendFrame(type: String, payload: Data) {
        var obj: [String: Any] = ["t": type, "p": payload.base64EncodedString()]
        if let d = deviceId { obj["d"] = d }
        guard let data = try? JSONSerialization.data(withJSONObject: obj) else { return }
        ws?.send(.string(String(decoding: data, as: UTF8.self))) { _ in }
    }

    private func teardown(_ error: Error?) {
        if let call = pairCall {
            call.reject("tunnel error: \(error?.localizedDescription ?? "closed")")
            pairCall = nil
        }
        ws?.cancel(with: .normalClosure, reason: nil)
        ws = nil
        urlSession = nil
        tx = nil
        rx = nil
        handshake = nil
    }

    // MARK: - device static keypair (Keychain-held X25519)

    private struct Identity {
        let privateKey: Curve25519.KeyAgreement.PrivateKey
        let publicKey: Data
    }

    private let keychainService = "dev.octo.mobile.tunnel"
    private let keychainAccount = "device_static_key"

    private func deviceIdentity() throws -> Identity {
        if let raw = keychainLoad() {
            let key = try Curve25519.KeyAgreement.PrivateKey(rawRepresentation: raw)
            return Identity(privateKey: key, publicKey: key.publicKey.rawRepresentation)
        }
        let key = Curve25519.KeyAgreement.PrivateKey()
        try keychainStore(key.rawRepresentation)
        return Identity(privateKey: key, publicKey: key.publicKey.rawRepresentation)
    }

    private func keychainLoad() -> Data? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: keychainAccount,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess else { return nil }
        return item as? Data
    }

    private func keychainStore(_ data: Data) throws {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: keychainAccount,
            kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
        ]
        SecItemDelete(query as CFDictionary)
        let status = SecItemAdd(query as CFDictionary, nil)
        if status != errSecSuccess {
            throw TunnelError.message("keychain store failed (\(status))")
        }
    }
}

private enum TunnelError: Error, LocalizedError {
    case message(String)
    var errorDescription: String? {
        if case .message(let m) = self { return m }
        return nil
    }
}

// MARK: - Noise XX (initiator), Noise_XX_25519_ChaChaPoly_SHA256, empty prologue
//
// A focused implementation of just the initiator's three steps — write e; read
// (e, ee, s, es); write (s, se) — rather than a general token interpreter, built
// on CryptoKit primitives. It is byte-compatible with the Go host's flynn/noise
// and the Android noise-java plugin.

private func hmacSHA256(key: Data, data: Data) -> Data {
    Data(HMAC<SHA256>.authenticationCode(for: data, using: SymmetricKey(data: key)))
}

// Noise HKDF: extract with the chaining key as salt, then expand into `outputs`
// chained blocks. Matches the spec's HKDF used by MixKey and Split.
private func noiseHKDF(chainingKey: Data, ikm: Data, outputs: Int) -> [Data] {
    let tempKey = hmacSHA256(key: chainingKey, data: ikm)
    let o1 = hmacSHA256(key: tempKey, data: Data([0x01]))
    let o2 = hmacSHA256(key: tempKey, data: o1 + Data([0x02]))
    if outputs == 2 { return [o1, o2] }
    let o3 = hmacSHA256(key: tempKey, data: o2 + Data([0x03]))
    return [o1, o2, o3]
}

// NoiseCipherState is one direction's transport cipher: a key plus a 64-bit
// nonce counter. The 96-bit ChaChaPoly nonce is 4 zero bytes then the counter
// little-endian — the Noise convention both other ends use.
private final class NoiseCipherState {
    private let key: SymmetricKey
    private var n: UInt64 = 0

    init(key: Data) { self.key = SymmetricKey(data: key) }

    private func nonce() -> ChaChaPoly.Nonce {
        var bytes = Data(count: 4)
        withUnsafeBytes(of: n.littleEndian) { bytes.append(contentsOf: $0) }
        return try! ChaChaPoly.Nonce(data: bytes)
    }

    func encrypt(_ plaintext: Data, ad: Data = Data()) throws -> Data {
        let box = try ChaChaPoly.seal(plaintext, using: key, nonce: nonce(), authenticating: ad)
        n &+= 1
        return box.ciphertext + box.tag
    }

    func decrypt(_ ciphertext: Data, ad: Data = Data()) throws -> Data {
        let tag = ciphertext.suffix(16)
        let ct = ciphertext.prefix(ciphertext.count - 16)
        let box = try ChaChaPoly.SealedBox(nonce: nonce(), ciphertext: ct, tag: tag)
        let pt = try ChaChaPoly.open(box, using: key, authenticating: ad)
        n &+= 1
        return pt
    }
}

private final class NoiseHandshakeXX {
    private var ck: Data
    private var h: Data
    private var k: SymmetricKey?
    private var n: UInt64 = 0

    private let staticPrivate: Curve25519.KeyAgreement.PrivateKey
    private let ephemeral: Curve25519.KeyAgreement.PrivateKey
    private var remoteEphemeral: Data?
    private(set) var remoteStatic: Data?

    init(staticPrivate: Curve25519.KeyAgreement.PrivateKey) throws {
        self.staticPrivate = staticPrivate
        self.ephemeral = Curve25519.KeyAgreement.PrivateKey()
        let name = Data("Noise_XX_25519_ChaChaPoly_SHA256".utf8) // exactly 32 bytes
        self.h = name.count <= 32 ? name + Data(count: 32 - name.count) : Data(SHA256.hash(data: name))
        self.ck = self.h
        // Noise Initialize() always MixHash(prologue). Ours is empty, but even
        // MixHash([]) advances h (= SHA256(h)); skipping it diverges the running
        // hash from the host and fails the first AEAD tag in msg2.
        mixHash(Data())
    }

    private func mixHash(_ data: Data) { h = Data(SHA256.hash(data: h + data)) }

    private func mixKey(_ ikm: Data) {
        let out = noiseHKDF(chainingKey: ck, ikm: ikm, outputs: 2)
        ck = out[0]
        k = SymmetricKey(data: out[1])
        n = 0
    }

    private func chaNonce() -> ChaChaPoly.Nonce {
        var bytes = Data(count: 4)
        withUnsafeBytes(of: n.littleEndian) { bytes.append(contentsOf: $0) }
        return try! ChaChaPoly.Nonce(data: bytes)
    }

    private func encryptAndHash(_ plaintext: Data) throws -> Data {
        guard let k = k else { mixHash(plaintext); return plaintext }
        let box = try ChaChaPoly.seal(plaintext, using: k, nonce: chaNonce(), authenticating: h)
        n &+= 1
        let ct = box.ciphertext + box.tag
        mixHash(ct)
        return ct
    }

    private func decryptAndHash(_ ciphertext: Data) throws -> Data {
        guard let k = k else { mixHash(ciphertext); return ciphertext }
        let tag = ciphertext.suffix(16)
        let ct = ciphertext.prefix(ciphertext.count - 16)
        let box = try ChaChaPoly.SealedBox(nonce: chaNonce(), ciphertext: ct, tag: tag)
        let pt = try ChaChaPoly.open(box, using: k, authenticating: h)
        n &+= 1
        mixHash(Data(ciphertext))
        return pt
    }

    private func dh(_ priv: Curve25519.KeyAgreement.PrivateKey, _ pubRaw: Data) throws -> Data {
        let pub = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: pubRaw)
        let secret = try priv.sharedSecretFromKeyAgreement(with: pub)
        return secret.withUnsafeBytes { Data($0) }
    }

    // msg1 (initiator → responder): e
    func writeMessage1() throws -> Data {
        let ePub = ephemeral.publicKey.rawRepresentation
        mixHash(ePub)
        let payload = try encryptAndHash(Data()) // no key yet → empty plaintext
        return ePub + payload
    }

    // msg2 read (e, ee, s, es) then msg3 write (s, se). Returns msg3 bytes.
    func readMessage2WriteMessage3(_ msg: Data) throws -> Data {
        var offset = msg.startIndex

        // e
        let re = msg.subdata(in: offset..<msg.index(offset, offsetBy: 32))
        offset = msg.index(offset, offsetBy: 32)
        remoteEphemeral = re
        mixHash(re)
        // ee
        mixKey(try dh(ephemeral, re))
        // s (encrypted static: 32 + 16 tag)
        let encStatic = msg.subdata(in: offset..<msg.index(offset, offsetBy: 48))
        offset = msg.index(offset, offsetBy: 48)
        let rs = try decryptAndHash(encStatic)
        remoteStatic = rs
        // es
        mixKey(try dh(ephemeral, rs))
        // payload (empty; 16-byte tag)
        let encPayload = msg.subdata(in: offset..<msg.endIndex)
        _ = try decryptAndHash(encPayload)

        // msg3: s, se
        let encMyStatic = try encryptAndHash(staticPrivate.publicKey.rawRepresentation)
        mixKey(try dh(staticPrivate, remoteEphemeral!))
        let encMsg3Payload = try encryptAndHash(Data())
        return encMyStatic + encMsg3Payload
    }

    // split derives the two transport keys. flynn/noise returns the
    // initiator→responder cipher first, so for the initiator send=out[0].
    func split() -> (send: Data, receive: Data) {
        let out = noiseHKDF(chainingKey: ck, ikm: Data(), outputs: 2)
        return (out[0], out[1])
    }
}

File: mobile/native/OctoTunnelPlugin.kt
package dev.octo.mobile

import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import android.util.Base64
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.southernstorm.noise.protocol.CipherState
import com.southernstorm.noise.protocol.DHState
import com.southernstorm.noise.protocol.HandshakeState
import com.southernstorm.noise.protocol.Noise
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONObject
import java.security.KeyStore
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

// OctoTunnel is the phone half of octo's managed tunnel — the native piece that
// cannot live in JavaScript. It holds the device's Noise static keypair in the
// Android Keystore, runs the Noise XX handshake with the host (brokered by the
// relay, which sees only ciphertext), and moves encrypted frames over a relay
// WebSocket. JavaScript hands it plaintext frames and receives plaintext frames
// back; no key material ever crosses the bridge.
//
// It mirrors the phone (initiator) side of the Go reference client in
// cmd/octo-relay/internal/client and interoperates with the host in
// internal/tunnel: Noise_XX_25519_ChaChaPoly_SHA256, empty prologue, JSON relay
// frames {t, tn, d, p} with p a base64 payload the relay never inspects.
@CapacitorPlugin(name = "OctoTunnel")
class OctoTunnelPlugin : Plugin() {

    // Noise suite / pattern shared with the host. Must match flynn/noise's
    // NewCipherSuite(DH25519, CipherChaChaPoly, HashSHA256) + HandshakeXX.
    private val protocolName = "Noise_XX_25519_ChaChaPoly_SHA256"

    private val http = OkHttpClient.Builder().build()

    // Session state. The WebSocket listener (one OkHttp thread) drives the
    // handshake and inbound decryption; send() runs on the Capacitor thread.
    // cryptoLock serializes cipher use — a Noise CipherState nonce is stateful.
    private val cryptoLock = Any()
    private var ws: WebSocket? = null
    private var hs: HandshakeState? = null
    private var tx: CipherState? = null
    private var rx: CipherState? = null
    private var deviceId: String? = null
    private var expectedHostKey: ByteArray? = null
    private var pairCall: PluginCall? = null

    // ── JS contract (native-bridge.ts) ──────────────────────────────────────

    @PluginMethod
    fun loadDeviceIdentity(call: PluginCall) {
        try {
            val (_, pub) = deviceKeypair()
            call.resolve(JSObject().put("publicKey", Base64.encodeToString(pub, Base64.NO_WRAP)))
        } catch (e: Exception) {
            call.reject("load identity failed: ${e.message}", e)
        }
    }

    @PluginMethod
    fun pair(call: PluginCall) {
        val relay = call.getString("relay")
        val token = call.getString("token")
        val hostKey = call.getString("hostKey")
        val tunnelId = call.getString("tunnelId")
        if (relay.isNullOrEmpty() || token.isNullOrEmpty() || hostKey.isNullOrEmpty()) {
            call.reject("pair: relay, token and hostKey are required")
            return
        }
        try {
            expectedHostKey = Base64.decode(hostKey, Base64.DEFAULT)

            // Build the initiator handshake with our persistent static keypair.
            val handshake = HandshakeState(protocolName, HandshakeState.INITIATOR)
            val (priv, _) = deviceKeypair()
            handshake.localKeyPair.setPrivateKey(priv, 0)
            handshake.start()
            hs = handshake
            tx = null
            rx = null
            deviceId = null
            pairCall = call

            // Dial the relay's device endpoint. OkHttp performs the WebSocket
            // upgrade over http(s); map ws(s):// accordingly.
            val base = relay
                .replaceFirst("wss://", "https://")
                .replaceFirst("ws://", "http://")
                .trimEnd('/')
            val url = deviceUrl(base, tunnelId, token)
            ws = http.newWebSocket(Request.Builder().url(url).build(), Listener())
        } catch (e: Exception) {
            pairCall = null
            call.reject("pair failed: ${e.message}", e)
        }
    }

    @PluginMethod
    fun send(call: PluginCall) {
        val frame = call.getString("frame")
        if (frame == null) {
            call.reject("send: frame is required")
            return
        }
        try {
            synchronized(cryptoLock) {
                val sender = tx ?: throw IllegalStateException("no active session")
                val pt = frame.toByteArray(Charsets.UTF_8)
                val out = ByteArray(pt.size + sender.macLength)
                val n = sender.encryptWithAd(null, pt, 0, out, 0, pt.size)
                sendFrame("data", out.copyOf(n))
            }
            call.resolve()
        } catch (e: Exception) {
            call.reject("send failed: ${e.message}", e)
        }
    }

    @PluginMethod
    fun disconnect(call: PluginCall) {
        teardown(null)
        call.resolve()
    }

    // ── relay WebSocket ─────────────────────────────────────────────────────

    private inner class Listener : WebSocketListener() {
        override fun onMessage(webSocket: WebSocket, text: String) {
            try {
                val f = JSONObject(text)
                when (f.optString("t")) {
                    "paired" -> {
                        deviceId = f.optString("d")
                        pump() // send handshake message 1
                    }
                    "handshake" -> {
                        val h = hs ?: return
                        val msg = Base64.decode(f.optString("p"), Base64.DEFAULT)
                        val payload = ByteArray(Noise.MAX_PACKET_LEN)
                        h.readMessage(msg, 0, msg.size, payload, 0)
                        pump()
                    }
                    "data" -> deliverData(f.optString("p"))
                }
            } catch (e: Exception) {
                teardown(e)
            }
        }

        override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
            teardown(t)
        }

        override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
            teardown(if (tx == null) IllegalStateException("relay closed before handshake") else null)
        }
    }

    // pump drives the initiator handshake: it writes every outgoing handshake
    // message the state machine is ready to send, then completes the session
    // (verify host key + split into transport ciphers) once the pattern is done.
    private fun pump() {
        val h = hs ?: return
        while (h.action == HandshakeState.WRITE_MESSAGE) {
            val buf = ByteArray(Noise.MAX_PACKET_LEN)
            val len = h.writeMessage(buf, 0, ByteArray(0), 0, 0)
            sendFrame("handshake", buf.copyOf(len))
        }
        if (h.action == HandshakeState.SPLIT) {
            completeHandshake(h)
        }
    }

    private fun completeHandshake(h: HandshakeState) {
        // Authenticate the host: its static key, learned during XX, must equal
        // the one the pairing QR carried. This is what the relay can't forge.
        val remote: DHState = h.remotePublicKey
        val hostPub = ByteArray(remote.publicKeyLength)
        remote.getPublicKey(hostPub, 0)
        val expected = expectedHostKey
        if (expected != null && !hostPub.contentEquals(expected)) {
            teardown(SecurityException("host key mismatch — refusing to trust relay peer"))
            return
        }
        val pair = h.split()
        synchronized(cryptoLock) {
            tx = pair.sender
            rx = pair.receiver
        }
        h.destroy()
        hs = null
        pairCall?.resolve()
        pairCall = null
    }

    private fun deliverData(payloadB64: String) {
        val text: String
        synchronized(cryptoLock) {
            val receiver = rx ?: return
            val ct = Base64.decode(payloadB64, Base64.DEFAULT)
            val out = ByteArray(ct.size)
            val n = receiver.decryptWithAd(null, ct, 0, out, 0, ct.size)
            text = String(out, 0, n, Charsets.UTF_8)
        }
        notifyListeners("frame", JSObject().put("frame", text))
    }

    private fun sendFrame(type: String, payload: ByteArray) {
        val o = JSONObject()
        o.put("t", type)
        deviceId?.let { o.put("d", it) }
        o.put("p", Base64.encodeToString(payload, Base64.NO_WRAP))
        ws?.send(o.toString())
    }

    private fun teardown(err: Throwable?) {
        pairCall?.let { call ->
            if (err != null) call.reject("tunnel error: ${err.message}", err as? Exception)
            pairCall = null
        }
        try {
            ws?.close(1000, null)
        } catch (_: Exception) {
        }
        ws = null
        synchronized(cryptoLock) {
            tx = null
            rx = null
        }
        hs?.destroy()
        hs = null
    }

    // ── device static keypair (Keystore-wrapped X25519) ─────────────────────
    //
    // AndroidKeyStore can't hold a raw X25519 scalar we can feed to Noise, so we
    // keep a Keystore-resident AES-GCM master key and store the 32-byte private
    // key wrapped by it. The private key is never persisted in the clear and the
    // master key never leaves the Keystore (hardware-backed where available).

    private val prefsName = "octo.tunnel"
    private val keyPriv = "device_priv" // base64(iv || ciphertext)
    private val keyPub = "device_pub" // base64
    private val masterAlias = "octo_tunnel_master"
    private val gcmTagBits = 128
    private val ivLen = 12

    // deviceUrl builds the relay device endpoint. Against a DNS-named relay it
    // prefixes the tunnel id as a subdomain (<tunnelId>.relay.octo.dev) so a
    // multi-node deployment's L4 balancer can consistent-hash the TLS SNI and
    // land the phone on the same node as its host. IP literals and dotless
    // hosts (local dev) dial unchanged. Mirrors relayDialBase in the host's
    // internal/tunnel/tunnel.go — keep the two rules identical.
    private fun deviceUrl(base: String, tunnelId: String?, token: String): String {
        val plain = "$base/device?token=$token"
        if (tunnelId.isNullOrEmpty()) return plain
        val uri = try { java.net.URI(base) } catch (_: Exception) { return plain }
        val host = uri.host ?: return plain
        val ipLike = host.matches(Regex("^[0-9.]+$")) || host.contains(":")
        if (!host.contains(".") || ipLike) return plain
        val portPart = if (uri.port != -1) ":${uri.port}" else ""
        return "${uri.scheme}://$tunnelId.$host$portPart/device?token=$token"
    }

    @Synchronized
    private fun deviceKeypair(): Pair<ByteArray, ByteArray> {
        val prefs = context.getSharedPreferences(prefsName, 0)
        val storedPriv = prefs.getString(keyPriv, null)
        val storedPub = prefs.getString(keyPub, null)
        if (storedPriv != null && storedPub != null) {
            return unwrap(Base64.decode(storedPriv, Base64.DEFAULT)) to Base64.decode(storedPub, Base64.DEFAULT)
        }

        val dh = Noise.createDH("25519")
        dh.generateKeyPair()
        val priv = ByteArray(dh.privateKeyLength)
        dh.getPrivateKey(priv, 0)
        val pub = ByteArray(dh.publicKeyLength)
        dh.getPublicKey(pub, 0)
        dh.destroy()

        prefs.edit()
            .putString(keyPriv, Base64.encodeToString(wrap(priv), Base64.NO_WRAP))
            .putString(keyPub, Base64.encodeToString(pub, Base64.NO_WRAP))
            .apply()
        return priv to pub
    }

    private fun masterKey(): SecretKey {
        val ks = KeyStore.getInstance("AndroidKeyStore").apply { load(null) }
        (ks.getKey(masterAlias, null) as? SecretKey)?.let { return it }
        val kg = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore")
        kg.init(
            KeyGenParameterSpec.Builder(
                masterAlias,
                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT,
            )
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .build(),
        )
        return kg.generateKey()
    }

    private fun wrap(plain: ByteArray): ByteArray {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, masterKey())
        val ct = cipher.doFinal(plain)
        return cipher.iv + ct
    }

    private fun unwrap(blob: ByteArray): ByteArray {
        val iv = blob.copyOfRange(0, ivLen)
        val ct = blob.copyOfRange(ivLen, blob.size)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, masterKey(), GCMParameterSpec(gcmTagBits, iv))
        return cipher.doFinal(ct)
    }
}
